### PR TITLE
Fix invalid point id retrieved in QVTKInteractorAdapter

### PR DESCRIPTION
**Pull request by Xaxetrov**
Added a mapping from qt to vtk ids.

### CODE REVIEW

**Comment by Xaxetrov** [comment]
@finetjul Created a very simple mapping, I don't know if it is enough...

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 1]
```diff
@@ -49,6 +49,7 @@ static const char* qt_key_to_key_sym(Qt::Key, Qt::KeyboardModifiers modifiers);
 
 QVTKInteractorAdapter::QVTKInteractorAdapter(QObject* parentObject)
   : QObject(parentObject), AccumulatedDelta(0), DevicePixelRatio(1)
+  , QtToVTKPointIds()
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: rename to `TouchPointIdMap`

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 2]
```diff
@@ -77,6 +78,30 @@ void QVTKInteractorAdapter::SetDevicePixelRatio(int ratio, vtkRenderWindowIntera
   }
 }
 
+void QVTKInteractorAdapter::UpdateQtToVTKPointIds(int qtPointId)
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: qtPointId ->touchPointId

Comment (finetjul) [reply GUISupport/Qt/QVTKInteractorAdapter.cxx]: should return the vtkPointId

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 3]
```diff
@@ -77,6 +78,30 @@ void QVTKInteractorAdapter::SetDevicePixelRatio(int ratio, vtkRenderWindowIntera
...
+    return;
+  }
+  int vtkPointId = 0;
+  bool inserted = false;
+  while (!inserted)
+  {
+    if (!QtToVTKPointIds.values().contains(vtkPointId))
+    {
+      QtToVTKPointIds.insert(qtPointId, vtkPointId);
+      inserted = true;
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: `break;`

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 4]
```diff
@@ -77,6 +78,30 @@ void QVTKInteractorAdapter::SetDevicePixelRatio(int ratio, vtkRenderWindowIntera
   }
 }
 
+void QVTKInteractorAdapter::UpdateQtToVTKPointIds(int qtPointId)
+{
+  if (QtToVTKPointIds.contains(qtPointId))
+  {
+    return;
+  }
+  int vtkPointId = 0;
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: `vtkPointId` -> `pointerIndex`

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 5]
```diff
@@ -187,9 +212,14 @@ bool QVTKInteractorAdapter::ProcessEvent(QEvent* e, vtkRenderWindowInteractor* i
       t == QEvent::TouchEnd)
   {
     QTouchEvent* e2 = dynamic_cast<QTouchEvent*>(e);
+    foreach(const QTouchEvent::TouchPoint& point, e2->touchPoints())
+    {
+      UpdateQtToVTKPointIds(point.id());
+    }
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: no need for a separate loop, merge it with below loop

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 6]
```diff
@@ -187,9 +212,14 @@ bool QVTKInteractorAdapter::ProcessEvent(QEvent* e, vtkRenderWindowInteractor* i
...
     foreach (const QTouchEvent::TouchPoint& point, e2->touchPoints())
     {
-      if (point.id() >= VTKI_MAX_POINTERS)
+      int vtkPointId = GetVTKPointId(point.id());
+      if (vtkPointId >= VTKI_MAX_POINTERS)
       {
         break;
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: don't break, instead, only process vtkPointId < VTKI_MAX_POINTERS

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 7]
```diff
@@ -198,18 +228,20 @@ bool QVTKInteractorAdapter::ProcessEvent(QEvent* e, vtkRenderWindowInteractor* i
...
-                                      0,0,0, point.id());
+                                      0,0,0, vtkPointId);
     }
     foreach (const QTouchEvent::TouchPoint& point, e2->touchPoints())
     {
-      if (point.id() >= VTKI_MAX_POINTERS)
+      int vtkPointId = GetVTKPointId(point.id());
+      if (vtkPointId >= VTKI_MAX_POINTERS)
       {
         break;
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: don't break to be sure to remove released points

File: GUISupport/Qt/QVTKInteractorAdapter.h [hunk 8]
```diff
@@ -72,6 +78,7 @@ class VTKGUISUPPORTQT_EXPORT QVTKInteractorAdapter : public QObject
 protected:
   int AccumulatedDelta;
   int DevicePixelRatio;
+  QMap<int, int> QtToVTKPointIds;
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.h]: I wonder if the following wouldn't be more elegant...
```
  int touchPointMap[VKTI_MAX_POINTER];
```

**Review by finetjul — CHANGES_REQUESTED** [review]

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 9]
```diff
@@ -77,6 +78,30 @@ void QVTKInteractorAdapter::SetDevicePixelRatio(int ratio, vtkRenderWindowIntera
   }
 }
 
+int QVTKInteractorAdapter::UpdateTouchPointIdMap(int qtPointId)
+{
+  if (TouchPointIdMap.contains(qtPointId))
+  {
+    return GetPointerIndex(qtPointId);
+  }
+  int pointerIndex = 0;
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: You could have optimize it. e.g.

```
int pointerIndex = this->TouchPointIdMap.value(qtPointId, -1);
if (pointerIndex != -1)
{
  return pointerIndex;
}
...
```

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 10]
```diff
@@ -77,6 +78,30 @@ void QVTKInteractorAdapter::SetDevicePixelRatio(int ratio, vtkRenderWindowIntera
...
+  for (; pointerIndex <= VTKI_MAX_POINTERS; ++pointerIndex)
+  {
+    if (!TouchPointIdMap.values().contains(pointerIndex))
+    {
+      TouchPointIdMap.insert(qtPointId, pointerIndex);
+      break;
+    }
+    pointerIndex++;
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: are you sure you want to increment pointerIndex twice ?

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 11]
```diff
@@ -77,6 +78,30 @@ void QVTKInteractorAdapter::SetDevicePixelRatio(int ratio, vtkRenderWindowIntera
...
+int QVTKInteractorAdapter::UpdateTouchPointIdMap(int qtPointId)
+{
+  if (TouchPointIdMap.contains(qtPointId))
+  {
+    return GetPointerIndex(qtPointId);
+  }
+  int pointerIndex = 0;
+  for (; pointerIndex <= VTKI_MAX_POINTERS; ++pointerIndex)
+  {
+    if (!TouchPointIdMap.values().contains(pointerIndex))
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]:
```
if (this->TouchPointIdMap.key(pointerIndex, -1) == -1)
{
  ...
}
```

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 12]
```diff
@@ -77,6 +78,30 @@ void QVTKInteractorAdapter::SetDevicePixelRatio(int ratio, vtkRenderWindowIntera
...
+    {
+      TouchPointIdMap.insert(qtPointId, pointerIndex);
+      break;
+    }
+    pointerIndex++;
+  }
+  return pointerIndex <= VTKI_MAX_POINTERS ? pointerIndex : -1;
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: test is not needed (pointerIndex is ALWAYS <= VTKI_MAX_POINTERS)

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 13]
```diff
@@ -77,6 +78,30 @@ void QVTKInteractorAdapter::SetDevicePixelRatio(int ratio, vtkRenderWindowIntera
...
+}
+
+int QVTKInteractorAdapter::GetPointerIndex(int qtPointId)
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: method can be const

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 14]
```diff
@@ -77,6 +78,30 @@ void QVTKInteractorAdapter::SetDevicePixelRatio(int ratio, vtkRenderWindowIntera
...
+{
+  return TouchPointIdMap[qtPointId];
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: use `value(qtPointId, -1)` it is safer

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 15]
```diff
@@ -187,9 +212,14 @@ bool QVTKInteractorAdapter::ProcessEvent(QEvent* e, vtkRenderWindowInteractor* i
       t == QEvent::TouchEnd)
   {
     QTouchEvent* e2 = dynamic_cast<QTouchEvent*>(e);
+    foreach(const QTouchEvent::TouchPoint& point, e2->touchPoints())
+    {
+      UpdateTouchPointIdMap(point.id());
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: this->UpdateTouchPointIdMap(...)

In VTK, we use `this->`

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 16]
```diff
@@ -187,9 +212,14 @@ bool QVTKInteractorAdapter::ProcessEvent(QEvent* e, vtkRenderWindowInteractor* i
...
+    }
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: Why didn't you merge it ?

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 17]
```diff
@@ -187,9 +212,14 @@ bool QVTKInteractorAdapter::ProcessEvent(QEvent* e, vtkRenderWindowInteractor* i
...
     foreach (const QTouchEvent::TouchPoint& point, e2->touchPoints())
     {
-      if (point.id() >= VTKI_MAX_POINTERS)
+      int pointerIndex = GetPointerIndex(point.id());
+      if (pointerIndex >= VTKI_MAX_POINTERS)
       {
         break;
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx]: what was wrong with my comment ?

**Review by finetjul — CHANGES_REQUESTED** [review]

File: GUISupport/Qt/QVTKInteractorAdapter.cxx (hunk 11 above)
Comment (Xaxetrov) [reply GUISupport/Qt/QVTKInteractorAdapter.cxx]: I cannot match your Qt knowledge^^

File: GUISupport/Qt/QVTKInteractorAdapter.cxx (hunk 12 above)
Comment (Xaxetrov) [reply GUISupport/Qt/QVTKInteractorAdapter.cxx]: It's a mistake, it should be <.
But pointerIndex will not be < VTKI_MAX_POINTERS if the map is full.

File: GUISupport/Qt/QVTKInteractorAdapter.cxx [hunk 18]
```diff
@@ -189,35 +213,36 @@ bool QVTKInteractorAdapter::ProcessEvent(QEvent* e, vtkRenderWindowInteractor* i
...
-      if (point.state() & Qt::TouchPointMoved)
+      int pointerIndex = this->GetPointerIndex(point.id());
+      if (pointerIndex < VTKI_MAX_POINTERS)
       {
-        iren->InvokeEvent(vtkCommand::MouseMoveEvent, nullptr);
+        iren->SetPointerIndex(pointerIndex);
+        if (point.state() & Qt::TouchPointReleased)
+        {
+          iren->InvokeEvent(vtkCommand::LeftButtonReleaseEvent, nullptr);
+          TouchPointIdMap.remove(point.id());
```
Comment (finetjul) [inline GUISupport/Qt/QVTKInteractorAdapter.cxx:236]: this->TouchPointIdMap

**Comment by finetjul** [comment]
Please update your commit message and say that the usage of touch point id was relying on IDs starting from 0 which is no longer the case

**Pull request closed without merging**